### ISSUE TITLE
Wrap() duplicates call stack

### ISSUE THREAD

**Issue by dc7303** [pkg/errors#242]
I am currently considering using pkg/errors.
However, I have a question, so I'm leaving a question.

I ran the [example code](https://github.com/pkg/errors/blob/5dd12d0cfe7f152f80558d591504ce685299311e/example_test.go#L115-L143) and expected the output shown in the code comments.
However, if you look at the output attached below, you can see that the output is completely different. The problem is that duplicate stacks are displayed.

And this also happens when I wrap the error passed to New() and WithStack(). Is this the intended behavior?

Environment:
pkg/errors version: 0.9.1
Go version: 1.16
OS version: macOS Big Sur 11.0.1

My output:
```bash
error
main.fn
        /Users/choedongcheol/Workspace/temp/test/main.go:100
main.ExampleWrap_extended
        /Users/choedongcheol/Workspace/temp/test/main.go:116
main.main
        /Users/choedongcheol/Workspace/temp/test/main.go:208
runtime.main
        /Users/choedongcheol/.gvm/gos/go1.16/src/runtime/proc.go:225
runtime.goexit
        /Users/choedongcheol/.gvm/gos/go1.16/src/runtime/asm_amd64.s:1371
inner
main.fn
        /Users/choedongcheol/Workspace/temp/test/main.go:101
main.ExampleWrap_extended
        /Users/choedongcheol/Workspace/temp/test/main.go:116
main.main
        /Users/choedongcheol/Workspace/temp/test/main.go:208
runtime.main
        /Users/choedongcheol/.gvm/gos/go1.16/src/runtime/proc.go:225
runtime.goexit
        /Users/choedongcheol/.gvm/gos/go1.16/src/runtime/asm_amd64.s:1371
middle
main.fn
        /Users/choedongcheol/Workspace/temp/test/main.go:102
main.ExampleWrap_extended
        /Users/choedongcheol/Workspace/temp/test/main.go:116
main.main
        /Users/choedongcheol/Workspace/temp/test/main.go:208
runtime.main
        /Users/choedongcheol/.gvm/gos/go1.16/src/runtime/proc.go:225
runtime.goexit
        /Users/choedongcheol/.gvm/gos/go1.16/src/runtime/asm_amd64.s:1371
outer
main.fn
        /Users/choedongcheol/Workspace/temp/test/main.go:103
main.ExampleWrap_extended
        /Users/choedongcheol/Workspace/temp/test/main.go:116
main.main
        /Users/choedongcheol/Workspace/temp/test/main.go:208
runtime.main
        /Users/choedongcheol/.gvm/gos/go1.16/src/runtime/proc.go:225
runtime.goexit
        /Users/choedongcheol/.gvm/gos/go1.16/src/runtime/asm_amd64.s:1371
```

**Comment by davecheney** [comment]
Yes, this is how it works. I couldn’t think of an algorithm to collapse related stack frames. I recommend not using WithStack at intermediary levels as it adds little value. WithStack was added because sometimes errors cross stack boudnaires, ie being lsssd through a channel to another goroutine.

**Comment by puellanivis** [comment]
This is a well known caveat with `pkg/errors`. You should try to ensure that at the earliest layer, all errors that are returned should be returned with a stack trace, and then each layer returning an error from a call into your own code uses `WithMessage`.

As davecheney points out, it’s really hard to algorithmically figure out when and where a stack trace should be attached, while for humans it’s a comparatively simple matter of policy enforcement.

**Comment by dc7303** [comment]
@davecheney @puellanivis Thank you so much for your reply.

**Comment by shubhamJay** [comment]
@davecheney @puellanivis , I understood `WithStack` shouldn't be used at intermediary levels. But would like to understand your thoughts on
1) Making `WithStack` function idempotent (by adding a condition to check if passed error is already a `withStack` type), so there may not be need of complex algo to collapse stack frames OR
2) Exposing a richer type to represent an error with stack (which I feel would be a more elegant way) so people do not end up calling `WithStack` multiple times.

Ps: I am very new to Golang, so please point out If I am missing out on something very obvious.

**Comment by puellanivis** [comment]
We already have as rich of an error type as is feasible to represent an error with a stack. It is not really feasible that functions should return any error type other than `error`. The ubiquity of the `error` type is an important part of ensuring ease of interoperability and consistency of code.

It is not always the right thing to do to never allow adding a secondary stacktrace. In 99% of cases, it never makes sense, but especially if a piece of code has a significant amount of error handling, a second stacktrace might actually make sense.

**Comment by shubhamJay** [comment]
Got it. Thank you @puellanivis.
